app/src/main/java/io/appium/settings/LocationTracker.java

```diff
@@ -0,0 +1,221 @@
+        } catch (SecurityException e) {
+            Log.e(TAG, "Appium Settings has no access to location permission");
+        }
+    }
+
+    private void stopLocationUpdatesWithPlayServices() {
+        if (mGoogleApiClient == null || !mGoogleApiClient.isConnected()) {
+            return;
+        }
+
+        Log.i(TAG, "Stopping Google Play Services location provider");
+        //noinspection deprecation
+        LocationServices.FusedLocationApi
+                .removeLocationUpdates(mGoogleApiClient, this);
+        mGoogleApiClient.disconnect();
+        mGoogleApiClient = null;
+    }
+
+    private void startLocationUpdatesWithoutPlayServices() {
+        if (mLocationManager == null) {
+            return;
+        }
+
+        boolean isGPSEnabled = mLocationManager
+                .isProviderEnabled(LocationManager.GPS_PROVIDER);
+        boolean isNetworkEnabled = mLocationManager
+                .isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+        if (!isGPSEnabled && !isNetworkEnabled) {
+            Log.e(TAG, "Both Coarse and GPS providers are disabled");
+            return;
+        }
+
+        if (isGPSEnabled) {
+            try {
+                mLocationManager.requestLocationUpdates(
+                        LocationManager.GPS_PROVIDER,
+                        LOCATION_UPDATES_INTERVAL,
+                        MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
+                locationProvider = LocationManager.GPS_PROVIDER;
+                Log.d(TAG, "GPS enabled. Getting FINE location");
+                return;
+            } catch (SecurityException e) {
+                Log.e(TAG, "Appium Settings has no access to FINE location permission");
+            }
+        }
+        try {
+            mLocationManager.requestLocationUpdates(
+                    LocationManager.NETWORK_PROVIDER,
+                    LOCATION_UPDATES_INTERVAL,
+                    MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
+            locationProvider = LocationManager.NETWORK_PROVIDER;
+            Log.d(TAG, "Network enabled. Getting COARSE location");
+        } catch (SecurityException e) {
+            Log.e(TAG, "Appium Settings has no access to COARSE location permission");
+        }
+    }
+
+    private void stopLocationUpdatesWithoutPlayServices() {
+        if (mLocationManager == null) {
+            return;
+        }
+
+        Log.i(TAG, "Stopping Android location provider");
+        mLocationManager.removeUpdates(this);
+        mLocationManager = null;
+    }
+    
+    @Nullable
+    public Location getLocation() {
+        if (mLocation == null) {
+            if (mGoogleApiClient != null) {
+                if (mGoogleApiClient.isConnected()) {
+                    try {
+                        //noinspection deprecation
+                        mLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleApiClient);
```

Comment:
Will we call `requestLocationUpdates` once and re-call `getLastLocation` again as a retry logic if it returns null?

https://stackoverflow.com/questions/29796436/why-is-fusedlocationapi-getlastlocation-null/29804454

I'm not sure how often the method returns null though...